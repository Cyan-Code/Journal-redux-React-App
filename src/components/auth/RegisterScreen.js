import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { uiRemoveError, uiSetError } from '../../actions/uiState';
import { startRegisterEmailPassword } from '../../actions/auth';

export const RegisterScreen = () => {

  const dispatch = useDispatch()
  
  const {msgError} = useSelector(state => state.ui) // Esto regresa un objeto, ( El objeto entero del estado de los dispatchs con los que interactuo )

  const [formValues, handleInputChange, ] = useForm({
    name: 'Luis',
    email: 'Luis@gmail',
    password: '123',
    password2: '123'
  });

  const {name, email, password, password2} = formValues;

  const handleRegister = (e) => {
    e.preventDefault();

    if( isFormValid() ){
      dispatch(startRegisterEmailPassword(email, password, name))
    }

  };

  const isFormValid = () => {

    if(name.trim().length === 0){
      dispatch(uiSetError('name is required'))
      return false;
    } else if (!validator.isEmail( email )) {
      dispatch(uiSetError('Email is not Valid'))
      return false
    } else if (password !== password2 || password.length < 5) {
      dispatch(uiSetError('password is wrong'))
      return false
    }
    dispatch( uiRemoveError() )
    return true;
  };

  return (
    <>
      <h3 className="auth__title">Register</h3>
        <form onSubmit = {handleRegister}>

          {
            msgError &&
            (
              <div className="auth__aler-error">
                {
                  msgError
                }
              </div>
            )
          }

          <input
            className="auth__input"
            type="text"
            placeholder="Name"
            autoComplete="off" 
            name="name"
            value = {name}
            onChange = { handleInputChange }
          />
          <input
            className="auth__input"
            type="text"
            placeholder="Email"
            name="email"
            value= {email}
            onChange = { handleInputChange }
          />
          <input
            className="auth__input"
            type="password"
            placeholder="Password"
            name="password"
            value= {password}
            onChange = { handleInputChange }
          />
          <input
            className="auth__input"
            type="password"
            placeholder="Confirm password"
            name="password2"
            value={password2}
            onChange = { handleInputChange }
          />
          <button
            className="btn btn-primary btn-block mb-5"
            type="submit"
          >
            Registered
          </button>

          <Link
            to="/auth/login"
            className="link"
          >
            Already registered?
          </Link>
        </form>
    </>
  )
}
