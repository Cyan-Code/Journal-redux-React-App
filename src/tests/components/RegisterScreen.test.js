import React from 'react';
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { mount } from "enzyme"
import { Provider } from "react-redux"
import { MemoryRouter } from "react-router-dom"
import { RegisterScreen } from "../../components/auth/RegisterScreen"
import { types } from '../../types/types';

const initState = {
  auth: {},
  ui:{
    loading: false,
    msgError: null
  }
}

const middlewares = [thunk] //estamos usando thunk
const mockStore = configureStore(middlewares)

let store = mockStore(initState)


const wrapper = mount(
  <MemoryRouter>
    <Provider store = { store }>
      <RegisterScreen/>
    </Provider>
  </MemoryRouter>
)

describe('Pruebas en el componente <RegisterScreen />', () => {
  
  test('Debe de mostrarse correctamente', () => {
    expect( wrapper ).toMatchSnapshot();
  })

  test('Debe de hacer el dispatch de la accion respectiva', () => {
    const emailFiled = wrapper.find('input[name="email"]')
    emailFiled.simulate('change', {
      target:{
        value:'',
        name:'email'
      }
    })
    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    })
    const actions = store.getActions()
    
    expect(actions[0]).toEqual({
      type: types.uiSetError,
      payload: 'Email is not Valid'
    })
  })

  test('Debe de mostrar la caja de lalerta con el error', () => {

    const initState2 = {
      auth: {},
      ui:{
        loading: false,
        msgError: 'Email incorrecto'
      }
    }
    const store2 = mockStore(initState2)
    const wrapper = mount(
      <MemoryRouter>
        <Provider store = { store2 }>
          <RegisterScreen/>
        </Provider>
      </MemoryRouter>
    );

    expect( wrapper.find('.auth__aler-error').exists() ).toBe(true)
    expect( wrapper.find('.auth__aler-error').text().trim() ).toBe(initState2.ui.msgError)
    
  })
  


})


