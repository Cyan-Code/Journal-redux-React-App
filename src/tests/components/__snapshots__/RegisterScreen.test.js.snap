// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <RegisterScreen /> Debe de mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Luis"
    />
    <input
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="Luis@gmail"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Registered
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
